e2e/aws/redshift_test.go

```diff
@@ -55,3 +65,221 @@ func testRedshiftServerless(t *testing.T) {
+			},
+		},
+		"auto user drop": {
+			user:   autoUserDrop,
+			dbUser: autoUserDrop,
+			query:  autoRolesQuery,
+			afterConnTestFn: func(t *testing.T) {
+				pgxConnMu.Lock()
+				defer pgxConnMu.Unlock()
+				waitForRedshiftAutoUserDrop(t, ctx, conn, autoUserDrop)
+			},
+		},
+	} {
+		test := test
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Run("connect", func(t *testing.T) {
+				route := tlsca.RouteToDatabase{
+					ServiceName: db.GetName(),
+					Protocol:    defaults.ProtocolPostgres,
+					Username:    test.dbUser,
+					Database:    "dev",
+				}
+				t.Run("via proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresConnTest(t, cluster, test.user, route, test.query)
+				})
+				t.Run("via local proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresLocalProxyConnTest(t, cluster, test.user, route, test.query)
+				})
+			})
+			if test.afterConnTestFn != nil {
+				test.afterConnTestFn(t)
+			}
+		})
+	}
+}
+
+func connectAsRedshiftClusterAdmin(t *testing.T, ctx context.Context, clusterID string) *pgx.Conn {
+	t.Helper()
+	info := getRedshiftAdminInfo(t, ctx, clusterID)
+	const dbName = "dev"
+	return connectAsPostgresAdmin(t, ctx, info, dbName)
+}
+
+func getRedshiftAdminInfo(t *testing.T, ctx context.Context, clusterID string) dbAdminInfo {
+	t.Helper()
+	cfg, err := config.LoadDefaultConfig(ctx,
+		config.WithRegion(mustGetEnv(t, awsRegionEnv)),
+	)
+	require.NoError(t, err)
+	clt := redshift.NewFromConfig(cfg)
+	result, err := clt.DescribeClusters(ctx, &redshift.DescribeClustersInput{
+		ClusterIdentifier: &clusterID,
+	})
+	require.NoError(t, err)
+	require.Len(t, result.Clusters, 1)
+	dbInstance := result.Clusters[0]
+	require.NotNil(t, dbInstance.MasterUsername)
+	require.NotNil(t, dbInstance.MasterPasswordSecretArn)
+	require.NotEmpty(t, *dbInstance.MasterUsername)
+	require.NotEmpty(t, *dbInstance.MasterPasswordSecretArn)
+	return dbAdminInfo{
+		address:  *dbInstance.Endpoint.Address,
+		port:     int(*dbInstance.Endpoint.Port),
+		username: *dbInstance.MasterUsername,
+		password: getMasterUserPassword(t, ctx, *dbInstance.MasterPasswordSecretArn),
+	}
+}
+
+// provisionRedshiftAutoUsersAdmin provisions an admin user suitable for auto-user
+// provisioning.
+func provisionRedshiftAutoUsersAdmin(t *testing.T, ctx context.Context, conn *pgx.Conn, adminUser string) {
+	t.Helper()
+	// Don't cleanup the db admin after, because test runs would interfere
+	// with each other.
+	_, _ = conn.Exec(ctx, fmt.Sprintf("CREATE USER %q WITH PASSWORD DISABLE", adminUser))
+	_, err := conn.Exec(ctx, fmt.Sprintf(`GRANT ROLE "sys:superuser" TO %q`, adminUser))
+	if err != nil {
+		require.ErrorContains(t, err, "already a member")
+	}
+}
+
+func waitForRedshiftAutoUserDeactivate(t *testing.T, ctx context.Context, conn *pgx.Conn, user string) {
+	t.Helper()
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		rows, _ := conn.Query(ctx, "SELECT 1 FROM pg_user_info as a WHERE a.usename = $1", user)
```

Comment:
Handle error?

Reply:
`conn.Query` documents this:

> // The returned Rows must be closed before the connection can be used again. It is safe to attempt to read from the
> // returned Rows even if an error is returned. The error will be the available in rows.Err() after rows are closed. It
> // **is allowed to ignore the error returned from Query and handle it in Rows.**
> //
> // Err() on the returned Rows must be checked after the Rows is closed to determine if the query executed successfully
> // as some errors can only be detected by reading the entire response. e.g. A divide by zero error on the last row.

So I intentionally check the error on the rows returned after I close them instead of the returned error. It's pretty odd to return an error if people generally shouldn't look at it  🤷‍♂️ 

I'll add a comment explaining this quirkiness 👍

Reply:
It's alright, feel free to leave as-is.

Reply:
just realized I need to do the same in rds as well, I'll update the error checking there to match this style